src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -670,215 +768,169 @@ object OrderedRVD {
+    fastKeys: Option[RDD[RegionValue]]
   ): OrderedRVD = coerce(
     typ,
     ContextRDD.weaken[RVDContext](rdd),
-    fastKeys.map(ContextRDD.weaken[RVDContext](_)),
-    hintPartitioner)
+    fastKeys.map(ContextRDD.weaken[RVDContext](_))
+    )
 
   def coerce(
     typ: OrderedRVDType,
     crdd: ContextRDD[RVDContext, RegionValue]
-  ): OrderedRVD = coerce(typ, crdd, None, None)
+  ): OrderedRVD = coerce(typ, crdd, None)
 
   def coerce(
     typ: OrderedRVDType,
     crdd: ContextRDD[RVDContext, RegionValue],
     fastKeys: ContextRDD[RVDContext, RegionValue]
-  ): OrderedRVD = coerce(typ, crdd, Some(fastKeys), None)
+  ): OrderedRVD = coerce(typ, crdd, Some(fastKeys))
+
+  def coerce(
+    typ: OrderedRVDType,
+    crdd: ContextRDD[RVDContext, RegionValue],
+    fastKeys: Option[ContextRDD[RVDContext, RegionValue]]
+  ): OrderedRVD = {
+    val keys = fastKeys.getOrElse(getKeys(typ, crdd))
+    makeCoercer(typ, keys).coerce(typ, crdd)
+  }
 
   def makeCoercer(
     fullType: OrderedRVDType,
-    crdd: ContextRDD[RVDContext, RegionValue],
-    hintPartitioner: Option[OrderedRVDPartitioner]): RVDCoercer = {
-    val sc = crdd.sparkContext
+    // keys: RDD[RegionValue[fullType.kType]]
+    keys: ContextRDD[RVDContext, RegionValue]
+   ): RVDCoercer = {
+    type CRDD = ContextRDD[RVDContext, RegionValue]
+    val sc = keys.sparkContext
 
     val emptyCoercer: RVDCoercer = new RVDCoercer(fullType) {
-      def _coerce(typ: OrderedRVDType, crdd: ContextRDD[RVDContext, RegionValue]): OrderedRVD = empty(sc, typ)
+      def _coerce(typ: OrderedRVDType, crdd: CRDD): OrderedRVD = empty(sc, typ)
     }
 
-    if (crdd.partitions.isEmpty)
-      return emptyCoercer
-
-    // keys: RDD[RegionValue[kType]]
-    val keys = crdd
-
-    val pkis = getPartitionKeyInfo(fullType, keys)
+    val pkis = getKeyInfo(fullType, keys)
 
     if (pkis.isEmpty)
       return emptyCoercer
 
-    val partitionsSorted = (pkis, pkis.tail).zipped.forall { case (p, pnext) =>
-      val r = fullType.pkType.ordering.lteq(p.max, pnext.min)
-      if (!r)
-        log.info(s"not sorted: p = $p, pnext = $pnext")
-      r
+    val bounds = pkis.map(_.interval).toFastIndexedSeq
+    val pkBounds = bounds.map(_.coarsen(fullType.partitionKey.length))
+    def orderPartitions = { crdd: CRDD =>
+      val pids = pkis.map(_.partitionIndex)
+      if (pids.isSorted && crdd.getNumPartitions == pids.length)
+        crdd
+      else
+        crdd.reorderPartitions(pids)
     }
+    val intraPartitionSortedness = pkis.map(_.sortedness).min
+
+    if (intraPartitionSortedness == OrderedRVPartitionInfo.KSORTED
+        && OrderedRVDPartitioner.isValid(fullType.kType, bounds)) {
 
-    val sortedness = pkis.map(_.sortedness).min
-    if (partitionsSorted && sortedness >= OrderedRVPartitionInfo.TSORTED) {
-      val (makeAdjustments, rangeBounds, adjSortedness) = rangesAndAdjustments(fullType, pkis, sortedness)
+      info("Coerced sorted dataset")
```

Comment:
For this path, we constraining to a partitioner that has the same right-end points as the existing one, the difference seems to be that we're actually enforcing a partition key.

Does this mean the big thing that might happen is that we'll collapse some partitions that would otherwise not satisfy the allowedOverlap property? Are there any tests / example of this behavior?

Reply:
Right, at most this is doing what the old `adjustPartitions` was doing: if we have one partition end at `[1, 2]`, and the next one start at `[1, 3]`, and we want a partition key length 1 (allowed overlap 0), then we should make the first partition end at `[1]` inclusive and the next one start at `[1]` exclusive, and the first rows of the second partition will need to be moved to the first.

Reply:
`OrderedRVDPartitionerSuite.testGenerateDisjoint` is the best source of examples of this.